### ISSUE TITLE
Bad certificate with UTFSTRING fields

### ISSUE THREAD

**Issue by ahmet2mir** [sensu/sensu-go#3985]
Hello,

We have an internal PKI creating certificates for sensu mTLS, but by default certificate is defined as:

```yaml
Country:             printablestring
State/County/Region: utf8string
Organization:        utf8string
CommonName:          utf8string
City/Locality:       utf8string
Organizational Unit: utf8string
emailAddress:        ia5string
```

And when starting sensu agent we got `remote error: tls: bad certificate`

If we force all fields to `printablestring`

```yaml
Country:             printablestring
State/County/Region: printablestring
Organization:        printablestring
CommonName:          printablestring
City/Locality:       printablestring
Organizational Unit: printablestring
emailAddress:        ia5string
```

Sensu start normally.

## Expected Behavior

Sensu must accept utf8string according to [rfc5280](https://www.ietf.org/rfc/rfc5280.txt)

```
X520CommonName ::= CHOICE {
      teletexString     TeletexString   (SIZE (1..ub-common-name)),
      printableString   PrintableString (SIZE (1..ub-common-name)),
      universalString   UniversalString (SIZE (1..ub-common-name)),
      utf8String        UTF8String      (SIZE (1..ub-common-name)),
      bmpString         BMPString       (SIZE (1..ub-common-name)) }


X520LocalityName ::= CHOICE {
      teletexString     TeletexString   (SIZE (1..ub-locality-name)),
      printableString   PrintableString (SIZE (1..ub-locality-name)),
      universalString   UniversalString (SIZE (1..ub-locality-name)),
      utf8String        UTF8String      (SIZE (1..ub-locality-name)),
      bmpString         BMPString       (SIZE (1..ub-locality-name)) }

X520StateOrProvinceName ::= CHOICE {
      teletexString     TeletexString   (SIZE (1..ub-state-name)),
      printableString   PrintableString (SIZE (1..ub-state-name)),
      universalString   UniversalString (SIZE (1..ub-state-name)),
      utf8String        UTF8String      (SIZE (1..ub-state-name)),
      bmpString         BMPString       (SIZE (1..ub-state-name)) }

X520OrganizationName ::= CHOICE {
      teletexString     TeletexString
                          (SIZE (1..ub-organization-name)),
      printableString   PrintableString
                          (SIZE (1..ub-organization-name)),
      universalString   UniversalString
                          (SIZE (1..ub-organization-name)),
      utf8String        UTF8String
                          (SIZE (1..ub-organization-name)),
      bmpString         BMPString
                          (SIZE (1..ub-organization-name))  }


X520OrganizationalUnitName ::= CHOICE {
      teletexString     TeletexString
                          (SIZE (1..ub-organizational-unit-name)),
      printableString   PrintableString
                          (SIZE (1..ub-organizational-unit-name)),
      universalString   UniversalString
                          (SIZE (1..ub-organizational-unit-name)),
      utf8String        UTF8String
                          (SIZE (1..ub-organizational-unit-name)),
      bmpString         BMPString
                          (SIZE (1..ub-organizational-unit-name)) }

X520countryName ::=     PrintableString (SIZE (2))
```

Extensions are all in OCTET STRING.

## Current Behavior

Sensu refuse to start

## Possible Solution

Issue is probably on CommonName.

## Steps to Reproduce (for bugs)

Example in ruby to set fields
```
require 'openssl'

def print_asn_strings(obj, depth = 0)
  if obj.respond_to? :each
    obj.each do |item|
      print_asn_strings(item, depth + 1)
    end
  else
    printf("%-40s %s\n", obj.value, obj.class) if (
      obj.class.to_s.match(/String/) &&
      obj.class != OpenSSL::ASN1::BitString
    )
  end
  nil
end

key = OpenSSL::PKey::RSA.new(4096)

subject = "/C=FR/ST=IDF/L=PARIS/O=Company/CN=myhost.example"

cert = OpenSSL::X509::Certificate.new

fields = []
OpenSSL::X509::Name.parse(subject).to_a.each do |field|
    fields << [field[0], field[1], OpenSSL::ASN1::UTF8STRING] # or OpenSSL::ASN1::PRINTABLESTRING
end

subject_x509 = OpenSSL::X509::Name.new(fields)

cert.subject = cert.issuer = subject_x509

cert.not_before = Time.now
cert.not_after = Time.now + 365*24*60*60
cert.public_key = key.public_key
cert.serial = 0x0
cert.version = 2

ef = OpenSSL::X509::ExtensionFactory.new
ef.subject_certificate = ef.issuer_certificate = cert

cert.add_extension ef.create_extension('basicConstraints', 'CA:FALSE', true)
cert.add_extension ef.create_extension('keyUsage', 'keyEncipherment,dataEncipherment,digitalSignature')
cert.add_extension ef.create_extension('subjectKeyIdentifier', 'hash')
cert.add_extension ef.create_extension('authorityKeyIdentifier', 'keyid:always,issuer:always')

cert.sign key, OpenSSL::Digest::SHA256.new

asn1 = OpenSSL::ASN1.decode(cert.to_der)

print_asn_strings(asn1)
```

## Context
Logs aren't very detailed, only "bad certificate" is shown.

## Your Environment

* sensuctl version 5.20.2+ee, enterprise edition, build c8d4472e852f36136536faf8b9fc4415d30a6137, built 2020-05-26T13:19:19Z, built with go1.13.7
* sensu-backend version 5.20.2+ee, enterprise edition, build c8d4472e852f36136536faf8b9fc4415d30a6137, built 2020-05-26T13:14:34Z, built with go1.13.7
* sensu-agent version 5.20.2+ee, enterprise edition, build c8d4472e852f36136536faf8b9fc4415d30a6137, built 2020-05-26T13:15:16Z, built with go1.13.7
* CentOS Linux release 7.8.2003 (Core)
* installation via RPM

**Comment by ccressent** [comment]
Bonjour Ahmet,

I've created a certificate with a CN set to "göögle", encoded as UTF8STRING in the ASN.1 (checked with `openssl asn1parse`), and used it in a test mTLS setup. The certificate was accepted and everything worked as expected.

We use Go's standard library to parse X509 certificates. It has been reported that it is stricter and less lenient than some other parsers. It can reject some certificates that would have been accepted by other, more lenient parsers (eg: in the PHP and Java world). In all cases of such rejection I've seen, the certificate was indeed not strictly correct upon close examination.

Would you be able to share one of the certificates that leads to the error? We can then inspect it with an ASN.1 decoder to see if something is slightly off with it. If the problematic certificate is a throwaway, testing one and you are able to share its private key, I can also try it in my test setup.

> Logs aren't very detailed, only "bad certificate" is shown.

Logs regarding this error are indeed not very detailed in 5.x; we've improved that in 6.0. Maybe you could try with a 6.0 test setup to see if you get a more precise error?

**Comment by ahmet2mir** [comment]
Sorry to not update this issue. 
I tested with a dummy mtls server following https://venilnoronha.io/a-step-by-step-guide-to-mtls-in-go with my certs.
And I got the same error "bad certficate".

Looks like Golang is strict in field checks, our CA is full printablestring where our cert tool create utf8.
Go refuse type mismatch (they probably don't strips off the outer sequence header).

So this is not a sensu issue, will close.
